docs/architecture-and-design.md

```diff
@@ -0,0 +1,390 @@
+  * [SSL-Gamecontroller](#ssl-gamecontroller)
+  * [grSim](#grsim)
+* [Important Classes](#important-classes)
+  * [World](#world)
+    * [Team](#team)
+    * [Robot](#robot)
+    * [Ball](#ball)
+    * [Field](#field)
+    * [Refbox and Gamestate](#refbox--gamestate)
+  * [Primitives](#primitives)
+  * [Intents](#intents)
+  * [Dynamic Parameters](#dynamic-parameters)
+  * [Robot Status](#robot-status)
+* [Design Patterns](#design-patterns)
+  * [Abstract Classes and Inheritance](#abstract-classes-and-inheritance)
+  * [Singleton Design Pattern](#singleton-design-pattern)
+  * [Factory Design Pattern](#factory-design-pattern)
+  * [Visitor Design Pattern](#visitor-design-pattern)
+  * [Observer Design Pattern](#observer-design-pattern)
+  * [C++ Templating](#c-templating)
+* [Coroutines](#coroutines)
+  * [What Are Coroutines?](#what-are-coroutines)
+  * [What Coroutines Do We Use?](#what-coroutines-do-we-use)
+  * [How Do We Use Coroutines?](#how-do-we-use-coroutines)
+* [Conventions](#conventions)
+  * [Coordinates](#coordinates)
+  * [Angles](#angles)
+* [Architecture Overview](#architecture-overview)
+  * [Diagram](#architecture-overview-diagram)
+  * [Backend](#backend)
+    * [Input](#input-responsibilities)
+    * [Output](#output-responsibilities)
+    * [Diagram](#backend-diagram)
+  * [AI](#ai)
+    * [Strategy](#strategy)
+      * [Skills / Actions](#skills--actions)
+      * [Tactics](#tactics)
+      * [Plays](#plays)
+    * [Navigation](#navigation)
+    * [Diagram](#ai-diagram)
+  * [Visualizer](#visualizer)
+
+
+# Tools
+A few commonly-used terms and tools to be familiar with:
+#### SSL-Vision
+  * This is the shared vision system used by the Small Size League. It is what connects to the cameras above the field, does the vision processing, and transmits the positional data of everything on the field to our AI computers.
+  * The GitHub repository can be found [here](https://github.com/RoboCup-SSL/ssl-vision)
+#### SSL-Gamecontroller
+  * Sometimes referred to as the "Refbox", this is another shared piece of Small Size League software that is used to send gamecontroller and referee commands to the teams. A human controls this application during the games to send the appropriate commands to the robots. For example, some of these commands are what stage the gameplay is in, such as `HALT`, `STOP`, `READY`, or `PLAY`.
+  * The GitHub repository can be found [here](https://github.com/RoboCup-SSL/ssl-game-controller)
+#### grSim
+  * The general robot simulator used by the Small-Size-League. We use this to manually test strategy since it is easy to place the robots and ball in desired locations, run a strategy, and see what the robots do. It is not perfectly accurate, but is useful for testing high-level logic.
+  * The GitHub repository can be found [here](https://github.com/RoboCup-SSL/grSim)
+
+
+# Important Classes
+These are classes that are either heavily used in our code, or are very important for understanding how the AI works.
```

Comment:
I noticed that `Action`, `Tactic`, and `Play` are not included in important classes even though they are being included in the AI section. Are they considered as important classes? If so, do you think a link at the end of this session to the AI session would be helpful?

Reply:
Yes, "Important Classes" were meant to be classes that are heavily used or important to know and _not_ part of the "core structure" of AI or other modules since those are easier to explain in their own sections. 

I updated the description of "Important Classes" to make this more clear. Let me know if that's better